lib/broadway_kafka/brod_client.ex

```diff
@@ -147,6 +153,20 @@ defmodule BroadwayKafka.BrodClient do
     end
   end
 
+  @impl true
+  def prepare_for_start(broadway_opts) do
+    {_, producer_opts} = broadway_opts[:producer][:module]
+    init_opts = Keyword.put(producer_opts, :broadway, broadway_opts)
+
+    case init(init_opts) do
+      {:error, message} ->
+        raise ArgumentError, "invalid options given to #{__MODULE__}.init/1, " <> message
+
+      {:ok, config} ->
+        {child_specs(config), broadway_opts}
```

Comment:
Instead of returning the options, should we return the initialized state?

Reply:
I don't think so. Since this callback is directly related to broadway's `prepare_for_start/2`, I think make sense they have the same return data.

IMO, the fact that our brod_client calls `init/1` here is an implementation detail that should not leak to the producer module. 

Do you see some advantages in returning the config data instead of broadway options?

If we return the initalized state here, how should we handle it on the the producer module? Gonna add it on the broadway options there? What you think?

Reply:
I think we should move this whole callback to the producer, and have the logic in the producer.

BrodClient is completely an implementation detail, it is only replaceable for tests, so I think going with the callback on the client and validating twice is adding extra indirection.

Reply:
Done @josevalim! 

Moved most of the logic from the brod_client (now its only returns the child_spec) to the producer and reuse the first initialization in order to avoid unnecessary validation.

Thanks for the feedback! :smile:  

Let me know if I can improve the solution in any other way!